composer/utils/object_store/s3_object_store.py

```diff
@@ -0,0 +1,161 @@
+            a config file or environment variables. Defaults to None.
+        client_config (dict, optional): Kwargs dictionary for :class:`botocore.config.Config`. Defaults to None.
+        transfer_config (dict, optional): Kwargs dictionary for :class:`boto3.s3.transfer.TransferConfig`. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        bucket: str,
+        region_name: Optional[str] = None,
+        endpoint_url: Optional[str] = None,
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        aws_session_token: Optional[str] = None,
+        client_config: Optional[Dict[Any, Any]] = None,
+        transfer_config: Optional[Dict[Any, Any]] = None,
+    ) -> None:
+        try:
+            import boto3
+            from boto3.s3.transfer import TransferConfig
+            from botocore.config import Config
+        except ImportError as e:
+            raise MissingConditionalImportError("s3", "boto3") from e
+        self.bucket = bucket
+        if client_config is None:
+            client_config = {}
+        config = Config(**client_config)
+        self.client = boto3.client(
+            's3',
+            config=config,
+            region_name=region_name,
+            endpoint_url=endpoint_url,
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
+            aws_session_token=aws_session_token,
+        )
+        if transfer_config is None:
+            transfer_config = {}
+        self.transfer_config = TransferConfig(**transfer_config)
+
+    def get_uri(self, object_name: str) -> str:
+        return f"s3://{self.bucket}/{object_name}"
+
+    def get_object_size(self, object_name: str) -> int:
+        try:
+            obj = self.client.get_object(Bucket=self.bucket, Key=object_name)
+        except Exception as e:
+            _ensure_not_found_errors_are_wrapped(self.get_uri(object_name), e)
+        return obj['ContentLength']
+
+    def upload_object(
+        self,
+        object_name: str,
+        filename: Union[str, pathlib.Path],
+        callback: Optional[Callable[[int, int], None]] = None,
+    ):
+        file_size = os.path.getsize(filename)
+        cb_wrapper = None if callback is None else lambda bytes_transferred: callback(bytes_transferred, file_size)
+        self.client.upload_file(Bucket=self.bucket,
+                                Key=object_name,
+                                Filename=filename,
+                                Callback=cb_wrapper,
+                                Config=self.transfer_config)
+
+    def download_object(
+        self,
+        object_name: str,
+        filename: Union[str, pathlib.Path],
+        overwrite: bool = False,
+        callback: Optional[Callable[[int, int], None]] = None,
+    ):
+        if os.path.exists(filename):
+            raise FileExistsError(f"The file at {filename} already exists")
```

Comment:
Curious, should this still occur if `overwrite=True`?

Reply:
Whoopsies, thanks.

Reply:
Fixed